documentation/src/algebraic-datatypes.md

```diff
@@ -0,0 +1,150 @@
+# Algebraic Datatypes
+
+Algebraic datatypes are also called discriminated unions, tagged unions, and sometimes "or types" colloquially. 
+They can be found in many functional languages (even if they're not named algebraic datatypes specifically), and have become widely known and used in Typescript 2.0.
+
+## Algebraic datatypes in C#
+
+While we don't have language support for algebraic datatypes in C#, there are some patterns on how you can implement something that behaves like one.
+
+We recommend the usage of the following pattern using an abstract base class with a private constructor and inner, deriving classes and usually a match function.
+This has the advantage that the algebraic options are namespaced as the inner class, which simplifies the naming.
+The constructor is private to prevent extension of the algebraic datatype - with a private constructor only inner classes can derive from the abstract base class.

```

Comment:
```suggestion
The constructor is private to prevent extension of the algebraic datatype — with a private constructor only inner classes can derive from the abstract base class.

```